cmake/caf-generate-enum-strings.cmake

```diff
@@ -0,0 +1,142 @@
+    string(REGEX REPLACE "^([0-9a-zA-Z_]+).*$" "\\1" tmp "${line}")
+    list(APPEND enum_values "${tmp}")
+  endif()
+endmacro()
+
+macro(write_enum_file)
+  if(is_enum_class)
+    set(label_prefix "${enum_name}::")
+  else()
+    set(label_prefix "")
+  endif()
+  string(REPLACE ";" "::" namespace_str "${namespaces}")
+  string(REPLACE "::" "/" namespace_path "${namespace_str}")
+  set(out "")
+  # File header and includes.
+  list(APPEND out
+    "#include \"caf/config.hpp\"\n"
+    "#include \"caf/string_view.hpp\"\n"
+    "\n"
+    "CAF_PUSH_DEPRECATED_WARNING\n"
+    "\n"
+    "#include \"${namespace_path}/${enum_name}.hpp\"\n"
+    "\n"
+    "#include <string>\n"
+    "\n"
+    "namespace ${namespace_str} {\n"
+    "\n")
+  # Generate to_string implementation.
+  list(APPEND out
+    "std::string to_string(${enum_name} x) {\n"
+    "  switch (x) {\n"
+    "    default:\n"
+    "      return \"???\"\;\n")
+  foreach(label IN LISTS enum_values)
+    list(APPEND out
+      "    case ${label_prefix}${label}:\n"
+      "      return \"${namespace_str}::${enum_name}::${label}\"\;\n")
+  endforeach()
+  list(APPEND out
+    "  }\n"
+     "}\n"
+     "\n")
+  # Generate from_string implementation.
+  list(APPEND out
+    "bool from_string(string_view in, ${enum_name}& out) {\n")
+  foreach(label IN LISTS enum_values)
+    list(APPEND out
+      "  if (in == \"${namespace_str}::${enum_name}::${label}\") {\n"
+      "    out = ${label_prefix}${label}\;\n"
+      "    return true\;\n"
+      "  }\n")
+  endforeach()
+  list(APPEND out
+    "  return false\;\n"
+    "}\n"
+    "\n")
+  # Generate from_integer implementation.
+  list(APPEND out
+    "bool from_integer(std::underlying_type_t<${enum_name}> in,\n"
+    "                  ${enum_name}& out) {\n"
+    "  auto result = static_cast<${enum_name}>(in)\;\n"
+    "  switch (result) {\n"
+    "    default:\n"
+    "      return false\;\n")
+  foreach(label IN LISTS enum_values)
+    list(APPEND out
+      "    case ${label_prefix}${label}:\n")
+  endforeach()
+  list(APPEND out
+    "      out = result\;\n"
+    "      return true\;\n"
+    "  }\n"
+    "}\n"
+    "\n")
+ # File footer.
+ list(APPEND out
+   "} // namespace ${namespace_str}\n"
+   "\n")
+ file(WRITE "${OUTPUT_FILE}" ${out})
+endmacro()
+
+function(main)
+  set(namespaces "")
+  set(enum_name "")
+  set(enum_values "")
+  set(is_enum_class FALSE)
+  file(STRINGS "${INPUT_FILE}" lines)
+  set(mode "seek")
+  foreach(line IN LISTS lines)
+    string(REGEX REPLACE "^(.+)(//.*)?" "\\1" line "${line}")
+    string(STRIP "${line}" line)
+    if(mode STREQUAL seek)
+      seek_mode()
+    elseif(mode STREQUAL scan)
+      scan_mode()
+    else()
+      # reached the end
+      if(NOT enum_name)
+        message(FATAL_ERROR "No enum found in input file")
+      endif()
+      if(NOT enum_values)
+        message(FATAL_ERROR "No enum values found for ${enum_name}")
+      endif()
+      write_enum_file()
```

Comment:
(Why is this not called `generate_mode`? Or why do the `_mode` functions not have similarly speaking names, e.g., `seek_enum` and `scan_cases`.)

Reply:
That's no longer a processing mode that processes lines. This step is writing the file and then returning from the function/script. But I did gave the two modes some better name.

Reply:
:)